File: 프로그래머스/lv3/이하은/섬 연결하기.kt
import java.util.*

data class Edge(
    val start: Int,
    val end: Int,
    val cost: Int
)

class Solution {
    lateinit var parent: MutableList<Int>
        
    fun solution(n: Int, costs: Array<IntArray>): Int {
        var answer = 0
        val edges = mutableListOf<Edge>()
        parent = MutableList(n + 1) { 0 }
        
        // 자기 자신을 부모 번호로 초기화 
        for(i in 1..n){
            parent[i] = i;
        }
        
        for(e in costs){
            edges.add(Edge(e[0], e[1], e[2]))
        }
        
        // 간선 비용이 작은 순으로 정렬 
        edges.sortBy{ edge -> edge.cost }
        
        // 사이클을 형성하지 않을 때만 MST에 포함 
        for(e in edges){
            if(findParent(e.start) != findParent(e.end)){
                unionParent(e.start, e.end)
                answer += e.cost
            }
        }
        
        return answer
    }
    
    fun findParent(x: Int): Int {
        if(x == parent[x]) return x
        parent[x] = findParent(parent[x])
        return parent[x]
    }

    fun unionParent(a: Int, b: Int){
        var a = findParent(a)
        var b = findParent(b)
        
        if(a < b) parent[b] = a
        else parent[a] = b
    }
}

File: 프로그래머스/lv3/장정윤/섬 연결하기.py
# 풀이 실패 -> 최소 거리: 크루스칼 알고리즘
def solution(n, costs):
    answer = 0
    uni = [0] * (n + 1)
    costs.sort(key = lambda x: x[2])
    for i in range(1,n + 1):
        uni[i] = i
    # union, find 분리된 집합 관리
    
    # 한 트리의 루트가 다른 트리 루트의 부모로 설정되어
    # 두 집합이 효과적으로 하나로 병합
    def union(x,y):
        a = find(x)
        b = find(y)
        if a != b:
            uni[a] = b
    # x와 루트로 가는 경로에 있는 모든 요소가 루트를 직접 가리키게 되어 트리 구조가 평면화
    # 원소가 속한 집합의 루트 결정
    def find(x):
        if uni[x] == x: return x
        else:
            uni[x] = find(uni[x])
            return uni[x]
    for n1,n2,cost in costs:
        if find(n1) != find(n2):
            union(n1,n2)
            answer += cost
    return answer